RavenDB-21874 - Can't add database ID to unused database IDs

shaharhikri opened the pull request:
### Issue link

https://issues.hibernatingrhinos.com/issue/RavenDB-21874/Cant-add-database-ID-to-unused-database-IDs

### Additional description

Can't add database ID to unused database IDs, it throws in base64string check.

### Type of change

- Bug fix


### How risky is the change?

- Low 

### Backward compatibility

- Non breaking change

### Is it platform specific issue?

- No

### Documentation update

- No documentation update is needed 

### Testing by Contributor

- It has been verified by manual testing

### Testing by RavenDB QA team

- No special testing by RavenDB QA team is needed

### Is there any existing behavior change of other features due to this change?

- No

### UI work

- No UI work is needed

shaharhikri committed:
RavenDB-21874 - Can't add database ID to unused database IDs

ayende commented on src/Raven.Server/Web/System/AdminDatabasesHandler.cs:
```diff
@@ -1309,8 +1309,13 @@ private async Task ValidateUnusedIdsAsync(HashSet<string> unusedIds, string data
+                base64 += "=";
+            
+            int base64Size = base64.Length;
+            var bytes = new byte[base64Size];
```
Note that you moved from stackalloc to a new array, intentional?
Also, what is the maximum size expected here?
Also, why do we even _do_ that for?

shaharhikri replied:
Done

haludi commented:
@shaharhikri, can you add a test?

ayende commented on src/Raven.Server/Web/System/AdminDatabasesHandler.cs:
```diff
@@ -1309,7 +1309,12 @@ private async Task ValidateUnusedIdsAsync(HashSet<string> unusedIds, string data
         {
-            int base64Size = (int)Math.Ceiling((double)base64.Length / 3) * 4;
+            // base64string length has to be a product of 4
+            var paddingLen = base64.Length%4;
+            for (int i = 0; i < paddingLen; i++)
+                base64 += "=";
```
You are now allocating, why? The whole point is to do this wihtout allocations

ppekrol replied:
Looks like DatabaseIds are not Base64 strings so the whole code can be removed, I do not see a point for having it

ayende commented on src/Raven.Server/Web/System/AdminDatabasesHandler.cs:
```diff
@@ -1309,7 +1309,12 @@ private async Task ValidateUnusedIdsAsync(HashSet<string> unusedIds, string data
+            for (int i = 0; i < paddingLen; i++)
+                base64 += "=";
+            
+            int base64Size = base64.Length;
```
Also, validate that this is small enough not to blow the stack

ppekrol replied:
this is not a hot path, you can use a ArrayPool

karmeli87 commented on test/SlowTests/Issues/RavenDB-21874.cs:
```diff
@@ -0,0 +1,47 @@
+            new HashSet<string> { "6ZY2cIMkCEOzFD3CtbdH1+" }, validate: true);
+
+        await store.Maintenance.Server.SendAsync(cmd);
+    }
```
let's also add a test that use an actual database id and succeed to remove it

shaharhikri replied:
Done

karmeli87 commented on src/Raven.Server/Web/System/AdminDatabasesHandler.cs:
```diff
@@ -1309,7 +1309,12 @@ private async Task ValidateUnusedIdsAsync(HashSet<string> unusedIds, string data
         public static unsafe bool IsBase64String(string base64)
         {
-            int base64Size = (int)Math.Ceiling((double)base64.Length / 3) * 4;
+            // base64string length has to be a product of 4
```
the thing is that `Base64Id` of the database is explicitly use only 22 chars
```
public string Base64Id { get; } = new string(' ', 22);
```
So let's make this method less generic
- check for 22 chars
- always add padding and validate

something like
```
 char* buffer = stackalloc char[24];
 fixed (char* str = dbId)
 {
     Buffer.MemoryCopy(str, buffer, 24 * sizeof(char), 22 * sizeof(char));
     buffer[22] = '=';
     buffer[23] = '=';
    return Convert.TryFromBase64String(buffer, ... );
 }
```

shaharhikri replied:
Done

karmeli87 commented on test/SlowTests/Issues/RavenDB-21874.cs:
```diff
@@ -0,0 +1,47 @@
+            new HashSet<string> { "6ZY2cIMkCEOzFD3CtbdH1@" }, validate: true); // @ is forbidden char
+
+        var e = await Assert.ThrowsAsync<RavenException>(() => store.Maintenance.Server.SendAsync(cmd));
+        Assert.True(e.InnerException is InvalidOperationException);
```
let's assert on the message here as well

shaharhikri replied:
Done

karmeli87 commented on test/SlowTests/Issues/RavenDB-21874.cs:
```diff
@@ -0,0 +1,47 @@
+
+namespace SlowTests.Issues;
+
+public class RavenDB_21874 : RavenTestBase
```
move this together with `ShouldValidateUnusedIds` test
also, in the `ShouldValidateUnusedIds` test assert for the error message

shaharhikri replied:
Done

shaharhikri committed:
RavenDB-21874 - Addressed PR Comments

ppekrol commented on src/Raven.Server/Web/System/AdminDatabasesHandler.cs:
```diff
@@ -1307,11 +1310,17 @@ private async Task ValidateUnusedIdsAsync(HashSet<string> unusedIds, string data
 
-        public static unsafe bool IsBase64String(string base64)
+        public static unsafe bool IsBase64Id(string base64)
         {
```
this should validate the length, it is expecting to get exactly 22 chars?

shaharhikri replied:
Yes, Done

ppekrol commented on src/Raven.Server/Web/System/AdminDatabasesHandler.cs:
```diff
@@ -1307,11 +1310,17 @@ private async Task ValidateUnusedIdsAsync(HashSet<string> unusedIds, string data
         }
 
-        public static unsafe bool IsBase64String(string base64)
+        public static unsafe bool IsBase64Id(string base64)
```
should be called IsDatabaseId? since this is not base64 actually?

shaharhikri replied:
Yes, I changed it

ppekrol commented on src/Raven.Server/Web/System/AdminDatabasesHandler.cs:
```diff
@@ -1265,8 +1265,11 @@ private async Task ValidateUnusedIdsAsync(HashSet<string> unusedIds, string data
+                if (id.Length != 22)
+                    throw new InvalidOperationException($"Database id '{id}' isn't valid because its length isn't 22.");
+
+                if (IsBase64Id(id)==false)
```
formatting

shaharhikri replied:
Done

ppekrol commented on src/Raven.Server/Web/System/AdminDatabasesHandler.cs:
```diff
@@ -1265,8 +1265,11 @@ private async Task ValidateUnusedIdsAsync(HashSet<string> unusedIds, string data
             {
-                if(IsBase64String(id)==false)
-                    throw new InvalidOperationException($"Database id '{id}' isn't valid because it isn't Base64String (it contains chars which cannot be in Base64String).");
+                if (id.Length != 22)
```
introduce a const and use it in both places, so there will be no surprises in the future if someone will change that

shaharhikri replied:
Done

shaharhikri committed:
RavenDB-21874 - Addressed PR Comments

ppekrol committed:
RavenDB-21874
- formatting
- use single const for the length
- move ValidateDatabaseId to the end of the handler for easier merging
- fixed typos, casing, string formatting
- cleanup

ppekrol commented:
Please review my changes

karmeli87 reviewed (approved):
LGTM
